File: trunk/src/client/ui/src/com/google/speedtracer/hintletengine/public/rules/hintlet_static_no_cookie.js
/*
 * Copyright 2009 Google Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

/**
 * Rule to find static content served from a domain that sets cookies.
 * 
 * Cloned from Page Speed: staticNoCookieLint.js
 */

// Make a namespace for this rule using a closure
(function() {  // Begin closure
var HINTLET_NAME = "Static Resource served from domains with cookies";

// Returns the cookie header if found, undefined if not found
var hasCookie = function(headers) {
  var cookie = hintlet.hasHeader(headers, 'Set-Cookie');
  if (cookie !== undefined && cookie.length > 0) {
    return cookie;
  }
  cookie = hintlet.hasHeader(headers, 'Cookie');
  if (cookie !== undefined && cookie.length > 0) {
    return cookie;
  }
  return undefined;
}

hintlet.register(HINTLET_NAME, function(dataRecord){
  if (dataRecord.type != hintlet.types.RESOURCE_FINISH) {
    return;
  }
  var resourceData = hintlet.getResourceData(dataRecord.data.identifier);
  if (!resourceData) {
    return;
  }

  var resourceType = hintlet.getResourceType(resourceData);

  // Make sure this is a static resource
  switch(resourceType) {
    case hintlet.RESOURCE_TYPE_STYLESHEET:
    case hintlet.RESOURCE_TYPE_SCRIPT:
    case hintlet.RESOURCE_TYPE_IMAGE:
    case hintlet.RESOURCE_TYPE_MEDIA:
      break;
    default:
      return;
  }

  var cookie = hasCookie(resourceData.responseHeaders);
  var sequence = dataRecord.sequence;

  if (cookie !== undefined) {
    hintlet.addHint(HINTLET_NAME, resourceData.responseReceivedTime,
        "URL " + resourceData.url + " is static content that should be "
        + "served from a domain that does not set cookies.  Found " 
        + (cookie.length + 8) + " extra bytes from cookie.",
        sequence, hintlet.SEVERITY_INFO);
  }
}); // End hintlet.register()

})();  // End closure

File: src/client/ui/src/com/google/speedtracer/hintletengine/client/rules/HintletStaticNoCookie.java
/*
 * Copyright 2011 Google Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.speedtracer.hintletengine.client.rules;

import com.google.speedtracer.client.model.EventRecord;
import com.google.speedtracer.client.model.HintRecord;
import com.google.speedtracer.client.model.NetworkResource;
import com.google.speedtracer.client.model.ResourceRecord;
import com.google.speedtracer.hintletengine.client.HintletHeaderUtils;
import com.google.speedtracer.hintletengine.client.HintletNetworkResources;
import com.google.speedtracer.hintletengine.client.WebInspectorType;
import com.google.speedtracer.shared.EventRecordType;

/**
 * Rule to find static content served from a domain that sets cookies. From Page Speed, "Static
 * content, such as images, JS and CSS files, don't need to be accompanied by cookies, as there is
 * no user interaction with these resources. You can decrease request latency by serving static
 * resources from a domain that doesn't serve cookies. This technique is especially useful for pages
 * referencing large volumes of rarely cached static content, such as frequently changing image
 * thumbnails, or infrequently accessed image archives. We recommend this technique for any page
 * that serves more than 5 static resources. (For pages that serve fewer resources than this, it's
 * not worth the cost of setting up an extra domain.)"
 */
public class HintletStaticNoCookie extends HintletRule {

  @Override
  public String getHintletName() {
    return "Static Resource served from domains with cookies";
  }

  @Override
  public void onEventRecord(EventRecord eventRecord) {

    if (eventRecord.getType() != EventRecordType.RESOURCE_FINISH) {
      return;
    }

    ResourceRecord resourceFinishEvent = eventRecord.cast();

    NetworkResource savedNetworkResource =
        HintletNetworkResources.getInstance().getResourceData(resourceFinishEvent.getIdentifier());
    if (savedNetworkResource == null) {
      return;
    }

    // Make sure this is a static resource
    WebInspectorType resourceType = WebInspectorType.getResourceType(savedNetworkResource);
    switch (resourceType) {
      case STYLESHEET:
      case SCRIPT:
      case IMAGE:
      case MEDIA:
        break;
      default:
        return;
    }

    String cookie = HintletHeaderUtils.hasCookie(savedNetworkResource.getResponseHeaders());
    if (cookie != null) {
      addHint(getHintletName(), savedNetworkResource.getResponseReceivedTime(), "URL "
          + savedNetworkResource.getUrl() + " is static content that should be "
          + "served from a domain that does not set cookies.  Found " + (cookie.length() + 8)
          + " extra bytes from cookie.", eventRecord.getSequence(), HintRecord.SEVERITY_INFO);
    }
  }
}
